### PR TITLE
Fix combined reductions with Kokkos::View

### PR DESCRIPTION
**Pull request by masterleinad**
Extracted from #4889. We should preserve the memory space in combined reductions if a View is given.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

**Review by PhilMiller — APPROVED** [review]

File: core/unit_test/TestReduce.hpp [hunk 1]
```diff
@@ -635,7 +635,7 @@ TEST(TEST_CATEGORY, int_combined_reduce_mixed) {
 
   uint64_t nsum = (nw / 2) * (nw + 1);
 
-  auto result1_v = Kokkos::View<int64_t, Kokkos::HostSpace>{"result1_v"};
+  auto result1_v = Kokkos::View<int64_t>{"result1_v"};
```
Comment (dalg24) [inline core/unit_test/TestReduce.hpp:638]: Why did you change the test.  This looks intentional.

Comment (crtrott) [reply core/unit_test/TestReduce.hpp:638]: It is intentional.

**Review by crtrott — CHANGES_REQUESTED** [review]

File: core/unit_test/TestReduce.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/TestReduce.hpp:638]: I changed it to demonstrate the issue that I encountered in #4889 that was failing with an incompatible View assignment error. Is there are reason to only allow `HostSpace` here?

Comment (crtrott) [reply core/unit_test/TestReduce.hpp:638]: Oh I see. Can we change the test to take four reduction args, i.e. result1_v with defaultspace, result2_v with HostSpace, result3 as a scalar, and a fourth using the reducers (which is wrapping a host view)?

**Review by crtrott — CHANGES_REQUESTED** [review]

**Comment by masterleinad** [comment]
It turned out that making this work for memory spaces that are not host-accessible is not that trivial. The combined reducer implementation basically forms a reducer that lives in host memory space. To construct the underlying values, the implementation so far tried to access the VIew via the call operator (in host execution space) if a View was given as reducer which obviously doesn't work if the View is allocated in non-host-accessible memory space.
Since the initial value isn't used anyway, I decided to default-construct one of the desired type instead of copying the one we have stored in the View. Since the value type in a View has to be default-constructible anyway (implicitly in the current implementation), that should work.
When the value stored in the combined reducer is copied back, we run into the same memory access problems and we just perform a `Kokkos::deep_copy` now if the memory space is not accessible from the host.

In the end, using View with non-host-accessible memory space in combined reductions is more expensive than a host-accessible View + deep_copy. Hence, we could also decide to restrict combined reductions to host-accessible memory spaces. Note that the original fix changing the memory space of the implicitly constructed view type is still required for #4889.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Only building the `OpenMPTarget-Clang` image failed.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 2]
```diff
@@ -294,13 +294,24 @@ struct CombinedReducerImpl<std::integer_sequence<size_t, Idxs...>, Space,
     return m_value_view;
   }
 
-  KOKKOS_FUNCTION
+  template <int Idx, class View>
+  constexpr static void write_one_value_back(
+      View const& view, typename View::value_type const& value) noexcept {
```
Comment (dalg24) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:299]:
```suggestion
      View const& view, typename View::const_value_type& value) noexcept {
```

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 3]
```diff
@@ -294,13 +294,24 @@ struct CombinedReducerImpl<std::integer_sequence<size_t, Idxs...>, Space,
...
+    if (Kokkos::SpaceAccessibility<typename View::memory_space,
+                                   Space>::assignable)
+      view() = value;
+    else
+      Kokkos::deep_copy(view, value);
```
Comment (dalg24) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:304]: Is that a new fencing bug when an execution space instance was passed?

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 4]
```diff
@@ -203,18 +203,18 @@ KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
 
 template <class T>
 KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
-    T&& arg) noexcept ->
+    T&&) noexcept ->
     typename std::enable_if<is_view<typename std::decay<T>::type>::value,
                             typename std::decay<T>::type>::type::value_type {
-  return arg();
+  return typename std::decay<T>::type::value_type{};
```
Comment (dalg24) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:209]: Plz resolve conflicts and prefer `std::decay_t` here and in the trailing return type

File: core/src/impl/Kokkos_Combined_Reducer.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:304]: Let me try to improve this a little.

File: core/src/impl/Kokkos_Combined_Reducer.hpp (hunk 4 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp]: Fixed.

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 5]
```diff
@@ -183,10 +183,9 @@ KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
 
 template <class T>
 KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
-    T&& arg) noexcept ->
-    typename std::enable_if_t<is_view<std::decay_t<T>>::value,
-                              std::decay_t<T>>::value_type {
-  return arg();
+    T&&) noexcept -> typename std::enable_if_t<is_view<std::decay_t<T>>::value,
```
Comment (dalg24) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:186]:
```suggestion
    T&&) noexcept -> std::enable_if_t<is_view<std::decay_t<T>>::value,
```

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 6]
```diff
@@ -266,13 +265,26 @@ struct CombinedReducerImpl<std::integer_sequence<size_t, Idxs...>, Space,
     return m_value_view;
   }
 
-  KOKKOS_FUNCTION
+  template <typename ExecutionSpace, int Idx, class View>
```
Comment (dalg24) [inline core/src/impl/Kokkos_Combined_Reducer.hpp:268]: Pick one, `typename` or `class`, but do not mix
```suggestion
  template <typename ExecutionSpace, int Idx, class View>
```

File: core/src/impl/Kokkos_Combined_Reducer.hpp (hunk 5 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:186]: No, `typename` is required here. We use the dependent type `value_type` for whatever `std::enable_if_t` gives us.

Comment (dalg24) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:186]: Sry

**Review by dalg24 — COMMENTED** [review]
Looks good now.  Can you think of a better name for `write_one_value_back`?

Did we ever discuss the semantics when mixing value and view?  Obviously now, if one of the argument is a view, then it is asynchronous.

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 7]
```diff
@@ -266,13 +265,26 @@ struct CombinedReducerImpl<std::integer_sequence<size_t, Idxs...>, Space,
     return m_value_view;
   }
 
-  KOKKOS_FUNCTION
+  template <class ExecutionSpace, int Idx, class View>
+  constexpr static void write_one_value_back(
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:269]: I only added it because the calling one was `constexpr`. Do you want me to drop it?

**Comment by masterleinad** [comment]
> Looks good now. Can you think of a better name for write_one_value_back?

I thought it's an appropriate name given that it's called by `write_value_back_to_original_references()`.

> Did we ever discuss the semantics when mixing value and view? Obviously now, if one of the argument is a view, then it is asynchronous.

I would think that we as a whole didn't spend much time talking about the semantics of combined reductions at all. I would rather say that we need to fence whenever one of the arguments is a value.

**Review by PhilMiller — APPROVED** [review]

**Review by dalg24 — COMMENTED** [review]
Sharing pending change I had.  Haven't reviewed the whole thing.

File: core/src/impl/Kokkos_Combined_Reducer.hpp [hunk 8]
```diff
@@ -183,18 +183,17 @@ KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
...
 
 template <class T>
 KOKKOS_INLINE_FUNCTION auto _get_value_from_combined_reducer_ctor_arg(
-    T&& arg) noexcept ->
+    T&&) noexcept ->
     typename std::enable_if_t<is_reducer<std::decay_t<T>>::value,
                               std::decay_t<T>>::value_type {
-  return arg.reference();
+  return typename std::decay_t<T>::value_type{};
 }
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/impl/Kokkos_Combined_Reducer.hpp]: See 78206f1

File: core/src/impl/Kokkos_Combined_Reducer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/impl/Kokkos_Combined_Reducer.hpp:263]: I'd say drop it.

And if the caller (I assume that is `write_value_back_to_original_references(...)`) is `constexpr`, that can't be correct now either, and should also have `constexpr` dropped.

**Review by nliber — DISMISSED** [review]

**Comment by masterleinad** [comment]
@crtrott @nliber *ping* I addressed your comments.

**Review by dalg24 — APPROVED** [review]

**Review by crtrott — APPROVED** [review]
